[1.18.2] Fix tags for custom forge registries

SizableShrimp opened the pull request:
This PR addresses a few tag-related regressions due to the 1.18.2 update:
* Allow custom forge registries to use tags again. A custom registry must call `RegistryBuilder#hasTags` to enable tags, otherwise __they are off by default__. 
* Allow querying a Holder based on ResourceKey, ResourceLocation, or value for tag-enabled registries
* Fix ForgeRegistryTagsProvider to use the right tag directory method
* Cleanup nullability annotations in `IForgeRegistry` and `ForgeRegistry` and move to jetbrains annotations
* Add custom forge registries to the root registry (global registries)
* Expose new `ITagManager` system for querying tag-related info from a forge registry
* Big change: **tagFolder() is gone. Modded registries must place their tags in `data/<namespace>/tags/<registryNamespace>/<registryPath>`** - For example with a registry name `examplemod:test_registry`, the tags folder may look like`data/mydatapackid/tags/examplemod/test_registry/`

Checking whether a custom registry entry is part of a tag can be done with something like so:
```java
    public boolean is(TagKey<CoolThing> key) {
        return coolThingForgeRegistry.tags().getTag(key).contains(this);
    }
```

Optional tags have been reintroduced with `ITagManager#createOptionalTagKey(ResourceLocation, Set<Supplier<V>>)`.

SizableShrimp committed:
Fix tags for custom forge registries

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Fix ForgeRegistryTagsProvider

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Add custom registries to root registry

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Remove tagFolder-related methods in RegistryBuilder

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

Cyborgmas commented:
You can probably delete the ForgeTagHandler class completely? Also if optional tags are no longer needed, it means the "default set" behaviour would be manual by modders, haven't tested but would imagine it requires checking if a tag is empty on the client and providing their own default behaviour. Not a big deal but thought it was worth mentioning.

The one problem I see is that you changed the tag folder logic for previously custom registries (i.e. enchantments) and mod added custom registries. For vanilla registries it's a bit forced as they don't have the "s" added, and I would say it's a bit too much to consider both folders to be valid. However, for custom registries, compatibility for the tag folder could be maintained and deprecated to be removed in 1.19 perhaps?

pupnewfster commented on src/main/java/net/minecraftforge/registries/IForgeRegistry.java:
```diff
@@ -34,17 +39,24 @@
+    @NotNull Holder<V> getHolderOrThrow(ResourceKey<V> key);
+    @NotNull HolderSet.Named<V> getOrCreateTag(TagKey<V> name);
+    @NotNull Optional<HolderSet.Named<V>> getTag(TagKey<V> name);
+    @NotNull Iterable<Holder<V>> getTagOrEmpty(TagKey<V> name);
```
I believe it may also make sense to expose the following methods that also exist on `MappedRegistry`:
- `boolean isKnownTagName(TagKey)`
- `Stream<Pair<TagKey<T>, HolderSet.Named<T>>> getTags()`
- `Stream<TagKey<T>> getTagNames()`

pupnewfster commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -167,12 +168,13 @@ public ResourceLocation getRegistryName()
+    public boolean supportsTags()
     {
-        return tagFolder;
+        return this.builder.getHasWrapper();
```
I feel like this potentially should just be stored in a local variable the same as everything else that references the builder.

pupnewfster commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -234,26 +236,87 @@ public ResourceLocation getKey(V value)
+    @NotNull
+    NamespacedHolderHelper<V> getHolderHelper()
+    {
+        if (!this.builder.getHasWrapper())
+            throw new IllegalStateException("Cannot query holder helper for non-wrapped registry");
+
+        return this.defaultKey != null
+                ? this.getSlaveMap(NamespacedDefaultedWrapper.Factory.ID, NamespacedDefaultedWrapper.class).getHolderHelper()
+                : this.getSlaveMap(NamespacedWrapper.Factory.ID, NamespacedWrapper.class).getHolderHelper();
```
This looks like it could be "simplified" (reduce the long slave map calls) to:
```java
Registry<V> wrapper = getWrapper();
if (wrapper == null)
    throw new IllegalStateException("Cannot query holder helper for non-wrapped registry");
return wrapper.getHolderHelper();
```

And then the suppress warnings for unchecked could probably be removed as well.

SizableShrimp replied:
This is not possible because the two slave maps have no common super type with getHolderHelper() declared

pupnewfster reviewed (commented):
I also agree with Cyborg that `ForgeTagHandler` can probably go as everything is commented out in it except for one method that is just returning an always empty set now.

I am not sure how worth it is keeping tag folders for custom registries until 1.19 as with 1.18.2 the tag changes are so breaking they will have to rewrite their things anyway.

pupnewfster commented:
One final thought is for:
> Expose API methods for holders and holder sets based on related Registry methods - only works for wrapped forge registries (all tag-enabled registries will have a wrapped Registry variant)

It likely makes sense to mention with a very short javadocs on the given methods that they will fail if the registry doesn't have a backing wrapped vanilla registry.

SizableShrimp commented:
> You can probably delete the ForgeTagHandler class completely? Also if optional tags are no longer needed, it means the "default set" behaviour would be manual by modders, haven't tested but would imagine it requires checking if a tag is empty on the client and providing their own default behaviour. Not a big deal but thought it was worth mentioning.

I am open to suggestions on being able to default a tag in the event it does not exist. Maybe this is possible with a simple change to NamespacedHolderHelper#bindTags.

> However, for custom registries, compatibility for the tag folder could be maintained and deprecated to be removed in 1.19 perhaps?

This is just unnecessary patching. Tags were already a breaking change for mods with TagKey so keeping this in is not worthwhile.

SizableShrimp committed:
Add javadocs and a few more methods from Registry

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Add back concept of optional tags

Provide more ways to retrieve Holder instances from IForgeRegistry

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Remove ForgeTagHandler and reimplement optional tags

Update DuplicateOptionalTagTest

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Fix compile error

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Fix filter check for ClientboundUpdateTagsPacket

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

pupnewfster commented on src/main/java/net/minecraftforge/registries/ForgeRegistries.java:
```diff
@@ -129,7 +129,7 @@
         //Forge
         public static final ResourceKey<Registry<DataSerializerEntry>> DATA_SERIALIZERS = key("data_serializers");
-        public static final ResourceKey<Registry<GlobalLootModifierSerializer<?>>> LOOT_MODIFIER_SERIALIZERS = key("forge:loot_modifier_serializers");
+        public static final ResourceKey<Registry<DataSerializerEntry>> LOOT_MODIFIER_SERIALIZERS = key("forge:loot_modifier_serializers");
```
Is this an intended change? It is hard to tell without being able to browse the code.

SizableShrimp replied:
No, this was likely a dumb IntelliJ refactor thing I missed. Resolved.

pupnewfster commented on src/main/java/net/minecraftforge/registries/NamespacedHolderHelper.java:
```diff
@@ -120,6 +133,11 @@ boolean isKnownTagName(TagKey<T> name)
 
+    void addOptionalTag(TagKey<T> name, Set<Supplier<T>> defaults)
+    {
+        this.optionalTags.computeIfAbsent(name, k -> new HashSet<>()).addAll(defaults);
```
I don't think this and the injecting of works quite right in terms of duplicates. As it is a set of suppliers it won't remove duplicates if there are two different instances supplying the same value. I think it may need a distinct call or something before the toList.

SizableShrimp committed:
Small cleanup and add `.distinct()` call

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Add `IForgeRegistry#holderByNameCodec`

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/DuplicateOptionalTagTest.java:
```diff
@@ -41,46 +42,47 @@
+       // if (!TAG_B.isDefaulted())
+       // {
+       //     LOGGER.warn("Second instance of optional tag is not defaulted!");
+       // }
```
If there are no plans at the moment to reintroduce the method to check if an optional tag is 'defaulted', this commented-out block should be removed.

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/NamespacedHolderHelper.java:
```diff
@@ -49,6 +52,7 @@
     private ObjectList<Holder.Reference<T>> holdersById = new ObjectArrayList<>(256);
     private Map<ResourceLocation, Holder.Reference<T>> holdersByName = new HashMap<>();
     private Map<T, Holder.Reference<T>> holders = new IdentityHashMap<>();
+    private Map<TagKey<T>, Set<Supplier<T>>> optionalTags = new IdentityHashMap<>();
```
The field can be declared `final`. Additionally, it may be nicer here to use a `Multimap<TagKey<T>, Supplier<T>>` instead. To retain the use of `IdentityHashMap`, use a multimap created through `Multimaps.newSetMultimap(new IdentityHashMap<>(), HashSet::new)`.

SizableShrimp replied:
Using a Multimap provides minimal cleanup, but I will take the suggestion.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/DuplicateOptionalTagTest.java:
```diff
@@ -41,46 +42,47 @@
+               .collect(Collectors.toUnmodifiableSet());
+       if (!tagAValues.equals(expected))
+       {
+           // TODO Turn this into an actual exception? This can easily get lost in logs if not looking carefully
```
Perhaps, we should include a `Throwable` parameter at the end of the logging call to print out a visible stacktrace, instead of throwing an exception which would halt loading?

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -234,26 +240,164 @@ public ResourceLocation getKey(V value)
+
+    @NotNull
+    @Override
+    public TagKey<V> createOptionalTagKey(ResourceLocation location, @NotNull Set<Supplier<V>> defaults)
+    {
+        TagKey<V> tagKey = createTagKey(location);
+
+        getHolderHelper().addOptionalTag(tagKey, defaults);
+
+        return tagKey;
+    }
```
This probably should be advertised when this PR is merged as the replacement for the old `createOptional` with defaults.

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -234,26 +240,164 @@ public ResourceLocation getKey(V value)
+    }
+
+    @Override
+    public @NotNull Holder<V> getHolderOrThrow(V value)
```
The location of this `@NotNull` annotation is different from the others.

sciwhiz12 commented on src/main/java/net/minecraftforge/network/filters/VanillaConnectionNetworkFilter.java:
```diff
@@ -42,8 +45,10 @@
 {
-
     private static final Logger LOGGER = LogManager.getLogger();
+    private static final Set<ResourceLocation> FORGE_ADDED_REGISTRIES = Stream.of(ForgeRegistries.Keys.DATA_SERIALIZERS, ForgeRegistries.Keys.LOOT_MODIFIER_SERIALIZERS, ForgeRegistries.Keys.WORLD_TYPES)
+            .map(ResourceKey::location)
+            .collect(Collectors.toUnmodifiableSet());
```
This set's existence should likely be noted in `ForgeRegistries`, for ease of maintenance should a new Forge-exclusive registry be added.

sciwhiz12 commented on src/main/java/net/minecraftforge/network/filters/VanillaConnectionNetworkFilter.java:
```diff
@@ -42,8 +45,10 @@
 public class VanillaConnectionNetworkFilter extends VanillaPacketFilter
 {
-
     private static final Logger LOGGER = LogManager.getLogger();
```
We should likely also change the use of `LogManager` with `LogUtils`.

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/RegistryManager.java:
```diff
@@ -133,9 +141,25 @@ public String getName()
     }
 
+    @SuppressWarnings("unchecked")
+    private <T> void registerRegistry(WritableRegistry<T> rootRegistry, ForgeRegistry<?> forgeReg) {
```
Formatting: braces on newlines.
```suggestion
    private <T> void registerRegistry(WritableRegistry<T> rootRegistry, ForgeRegistry<?> forgeReg)
    {
```

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/RegistryManager.java:
```diff
@@ -133,9 +141,25 @@ public String getName()
+            rootRegistry.register((ResourceKey<T>) forgeReg.getRegistryKey(), (T) wrapper, Lifecycle.experimental());
+    }
+
+    public static CompletableFuture<List<Throwable>> preNewRegistryEvent(final Executor executor, final IModStateTransition.EventGenerator<? extends RegistryEvent.NewRegistry> eventGenerator) {
```
Formatting: braces on newlines.
```suggestion
    public static CompletableFuture<List<Throwable>> preNewRegistryEvent(final Executor executor, final IModStateTransition.EventGenerator<? extends RegistryEvent.NewRegistry> eventGenerator)
    {
```

SizableShrimp committed:
Cleanup with suggestions from sciwhiz12

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Add package-private `IHolderHelperHolder` interface

Cleans up `ForgeRegistry#getHolderHelper` implementation

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Fix folder generation for `TagsProvider`

This would previously crash for BuiltinRegistries
ForgeRegistryTagsProvider also unintentionally prefixed tags/ to the folder and resulted in "tags/tags/"

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Add forge-expanded constructors for biome and configured structure tag providers

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

sciwhiz12 commented on patches/minecraft/net/minecraft/data/tags/BiomeTagsProvider.java.patch:
```diff
@@ -0,0 +1,15 @@
+ import net.minecraft.world.level.biome.Biomes;
+ 
+ public class BiomeTagsProvider extends TagsProvider<Biome> {
++   @Deprecated
```
This should have a javadoc with `@deprecated` directing users to the other constructor.

sciwhiz12 commented on patches/minecraft/net/minecraft/data/tags/ConfiguredStructureTagsProvider.java.patch:
```diff
@@ -0,0 +1,15 @@
+ import net.minecraft.world.level.levelgen.structure.BuiltinStructures;
+ 
+ public class ConfiguredStructureTagsProvider extends TagsProvider<ConfiguredStructureFeature<?, ?>> {
++   @Deprecated
```
This should have a javadoc with `@deprecated` directing users to the other constructor.

sciwhiz12 commented on patches/minecraft/net/minecraft/data/tags/BiomeTagsProvider.java.patch:
```diff
@@ -0,0 +1,15 @@
+    public BiomeTagsProvider(DataGenerator p_211094_) {
+       super(p_211094_, BuiltinRegistries.f_123865_);
++   }
++   public BiomeTagsProvider(DataGenerator p_211094_, String modId, @javax.annotation.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
```
Should we be using the `javax.annotation` annotation here, or the JetBrains one? Also, the `forge.exc` file needs to be updated to add this.

sciwhiz12 commented on patches/minecraft/net/minecraft/data/tags/ConfiguredStructureTagsProvider.java.patch:
```diff
@@ -0,0 +1,15 @@
+    public ConfiguredStructureTagsProvider(DataGenerator p_211098_) {
+       super(p_211098_, BuiltinRegistries.f_123862_);
++   }
++   public ConfiguredStructureTagsProvider(DataGenerator p_211098_, String modId, @javax.annotation.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
```
Should we be using the `javax.annotation` annotation here, or the JetBrains one? Also, the `forge.exc` file needs to be updated to add this.

sciwhiz12 commented on patches/minecraft/net/minecraft/data/tags/GameEventTagsProvider.java.patch:
```diff
@@ -8,8 +8,8 @@
-+   public GameEventTagsProvider(DataGenerator p_126517_, String modId, @javax.annotation.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
-+      super(p_126517_, Registry.f_175412_, modId, existingFileHelper);
++   public GameEventTagsProvider(DataGenerator p_176826_, String modId, @javax.annotation.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
++      super(p_176826_, Registry.f_175412_, modId, existingFileHelper);
```
The `forge.exc` file needs to be updated with this change in SRG parameter name.

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/IHolderHelperHolder.java:
```diff
@@ -0,0 +1,6 @@
+package net.minecraftforge.registries;
```
Missing license header, please run the `updateLicenses`/`licenseFormat` task.

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -234,26 +240,165 @@ public ResourceLocation getKey(V value)
+    NamespacedHolderHelper<V> getHolderHelper()
+    {
+        Registry<V> wrapper = getWrapper();
+        if (!(wrapper instanceof IHolderHelperHolder))
```
As this cannot be a pattern-matching `instanceof` as my local testing reveals, this should be noted as a comment so future maintainers know at a glance why this does not use pattern-matching `instanceof`.

SizableShrimp committed:
More cleanup suggested by sciwhiz12

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Prefix non-vanilla registry tag dirs w/ namespace

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Prevent duplicate registry name registration

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Extract out ITagManager system

Designed with suggestions from Lex

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Ensure you can't remove tags from iterator

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Add `/forge tags` command

Co-authored-by: sciwhiz12 <arnoldnunag12@gmail.com>
Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Rename `IForgeRegistry#getTagManager` to `tags`

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Clarify a few nullability annotations

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

pupnewfster commented on src/main/java/net/minecraftforge/registries/tags/IReverseTag.java:
```diff
@@ -0,0 +1,22 @@
+
+public interface IReverseTag<V>
+{
+    Stream<TagKey<V>> getTagKeyStream();
```
As there is no other getter and it is obvious this is a stream due to the return type I feel like a better name for this may just be `getTagKeys`

pupnewfster commented on src/main/java/net/minecraftforge/registries/tags/ITagManager.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    @NotNull ITag<V> getTag(@NotNull TagKey<V> name);
+
+    @NotNull Optional<IReverseTag<V>> getReverseTag(@NotNull V value);
```
I feel like there should be some more javadocs on this interface, especially for example on this method giving a brief explanation of what a "reverse tag" even is (or maybe better yet have `IReverseTag` and `ITag` have javadocs both on the class level and method level.

pupnewfster commented on src/main/java/net/minecraftforge/registries/tags/ITagManager.java:
```diff
@@ -0,0 +1,45 @@
+     * Creates a tag key that will use the set of defaults if no tag JSON is found.
+     * Useful on the client side when a server may not provide a specific tag.
+     */
+    @NotNull TagKey<V> createOptionalTagKey(@NotNull ResourceLocation location, @NotNull Set<Supplier<V>> defaults);
```
Would there potentially be a slight benefit to expanding the set's generic to `Set<? extends Supplier<V>>`? That way if someone is passing a set of say registry objects they don't have to worry about explicitly declaring the set as being of suppliers?

pupnewfster commented on src/main/java/net/minecraftforge/registries/ForgeRegistryTagManager.java:
```diff
@@ -0,0 +1,133 @@
+
+        holderTags.forEach((key, holderSet) -> {
+            bound.add(key);
+            ((ForgeRegistryTag<V>) newTags.computeIfAbsent(key, ForgeRegistryTag::new)).bind(holderSet);
```
Rather than having this cast could you just have the value type of `newTags` be `ForgeRegistryTag<V>`? and then just have `tags` be `Map<TagKey<V>, ? extends ITag<V>>` so that you can set it to `newTags` without any casting trickery.

SizableShrimp replied:
This does not seem to work.

pupnewfster replied:
Ahh you are right, I missed `newTags` taking `tags` as a parameter of defaults initially when I suggested it

pupnewfster commented on src/main/java/net/minecraftforge/registries/ForgeRegistryTagManager.java:
```diff
@@ -0,0 +1,133 @@
+
+        this.owner.getHolderHelper().ifPresent(h -> h.addOptionalTag(tagKey, defaults));
+
+        return tagKey;
```
Part of me wonders if this method should have some sort of "failure" return for if it was unable to add an optional tag (such as if there is no holder helper), rather than just returning a new tag key. (This might not make sense to do, just figured I would ask as it did cross my mind)

SizableShrimp replied:
Failure return or failure exception?

pupnewfster replied:
Not sure of what the best would be. Maybe an optional or null? Might get a bit messy or even maybe an exception though that might not make sense. This is in part why I was mentioning it as I wonder if it makes sense to do something here, but I don't know *what*.

pupnewfster commented on src/main/java/net/minecraftforge/server/command/TagsCommand.java:
```diff
@@ -0,0 +1,266 @@
+         * /forge tags <registry> get <tag> [page]
+         * /forge tags <registry> query <element> [page]
+         */
+        return Commands.literal("tags")
```
I don't see a major reason not to allow this, but do we want this command to be accessible by everyone by default (like it is now)? Or do we want to potentially require the sender to be an op

pupnewfster commented on src/main/java/net/minecraftforge/server/command/TagsCommand.java:
```diff
@@ -0,0 +1,266 @@
+        final long tagCount = registry.getTags().count();
+
+        ctx.getSource().sendSuccess(createMessage(
+                new TranslatableComponent("%s", new TextComponent(registryKey.location().toString()).withStyle(ChatFormatting.GOLD)),
```
What is with the random TranslatableTextComponent of just `%s`? Why doesn't this just pass the created `TextComponent` directly and then change `createMessage` to take a `MutableComponent` instead of a `TranslatableComponent`

SizableShrimp replied:
@sciwhiz12

sciwhiz12 replied:
This was a remnant of when I refactored the common code between `listTags` and `listTagElements` into the separate helper method `createMessage`. As can be seen in the other methods, `TranslatableComponent` is used because of `%s / %s` for listing tag elements and `%s : %s` for listing tags which hold a certain element.

Additionally, the use of `TranslatableComponent` allows flexibility in changing the format of the command's output at either the programmer's discretion (by changing the string literal) or the final translator's discretion (through the translation keys), which would not otherwise be attainable if passing the text component directly.

pupnewfster replied:
A couple counter points to your mention of flexibility:
1. The programmer can always wrap it in a translatable component if they want to change it further in the future, that isn't a great reason to add an extra wrapper around it in the meantime.
2. This is a more valid reason, except that the key isn't exposed as a lang entry? So there is no way for the final translator to adjust it; at which point having it doesn't gain anything.

pupnewfster commented on src/main/java/net/minecraftforge/server/command/TagsCommand.java:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <I, O> O cast(final I input)
```
Is the `I` being a generic actually needed? Or can this just be `private static <O> cast(final Object input)`

SizableShrimp committed:
Expand documentation of tag manager system

Fix some points made by pupnewfster

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

pupnewfster commented on patches/minecraft/net/minecraft/data/tags/TagsProvider.java.patch:
```diff
@@ -9,21 +9,25 @@
 +      this(p_126546_, p_126547_, modId, existingFileHelper, null);
 +   }
-+   protected TagsProvider(DataGenerator p_126546_, Registry<T> p_126547_, String modId, @javax.annotation.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper, @javax.annotation.Nullable String folder) {
++   protected TagsProvider(DataGenerator p_126546_, Registry<T> p_126547_, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper, @org.jetbrains.annotations.Nullable String folder) {
```
Is this constructor with the folder being passed in just being kept for backward compat? As it seems to be unneeded now that `TagManager.getTagDir` exists and supports custom registries properly as of this PR? If so it should either be removed (as tags got broken so much and it isn't worth maintaining until 1.19 as people will have to adjust anyway) or it should be deprecated and the one without the folder passed in be preferred.

In fact I think what probably should happen is:
- remove this constructor (just have the one without the folder passed in)
- remove the class level `folder` variable as it isn't needed anymore
- Don't substring the `tags/` out of the returned folder from `getTagDir` and just pass it directly to the `ResourceType` constructor without prefixing it with `tags/` there in case any tag directories ever end up in a different subfolder

pupnewfster commented on src/main/java/net/minecraftforge/registries/ForgeRegistries.java:
```diff
@@ -127,7 +127,8 @@
-        //Forge
+        // Forge
+        // New forge-added registries should be populated in VanillaConnectionNetworkFilter#FORGE_ADDED_REGISTRIES
         public static final ResourceKey<Registry<DataSerializerEntry>> DATA_SERIALIZERS = key("data_serializers");
```
Is there a reason this isn't in forge's domain? If not it may make sense to add a TODO for 1.19 to move it to forge's domain and remove having to add them to `FORGE_ADDED_REGISTRIES` and instead just check if the namespace is forge's? (Or would that set still make sense to have for purposes if forge ever decides to add any registries itself that supports tags)

SizableShrimp replied:
I cannot comment on why this isn't in forge's domain. However, I would rather hardcode and provide a nice reference directly above the relevant code rather than rely on the namespace.

pupnewfster commented on src/main/java/net/minecraftforge/registries/RegistryManager.java:
```diff
@@ -133,9 +144,28 @@ public String getName()
+            final IModStateTransition.EventGenerator<? extends RegistryEvent.NewRegistry> eventGenerator)
+    {
+        return CompletableFuture.runAsync(() -> {
+            registerToRoot = true;
```
This could use a brief comment explaining why it isn't just "always" true as it is not that obvious from looking at the code that it is because we don't want to register any of the registries that get staticly initialized in `GameData` to root.

SizableShrimp replied:
This is a good point. The real reason is that we don't want to try and register vanilla-wrapped registries twice. As Registry is patched with all the forge vanilla-wrapped registries which queries GameData, and then returns it back to vanilla to be registered to the root registry. So the RegistryManager will try to register all the GameData registries to the root registry first, and then the vanilla Registry patches will attempt it as well and crash. We could maybe instead simplify this by always registering to root and changing the patches in Registry to return the instance but *not* register it to the root registry. Thoughts?

pupnewfster commented on src/main/java/net/minecraftforge/registries/tags/ITag.java:
```diff
@@ -0,0 +1,30 @@
+ * A tag is a collection of elements with an identifying {@link TagKey tag key}.
+ * For Forge, these are bound on world load.
+ * Tags will always be empty until they are bound.
+ * A tag instance provided for a given {@link TagKey} from a given {@link ITagManager} will always return the same instance on future invocations.
```
Is this statement of same instance also true for if a reload occurs? Either way it is probably worth clarifying it.

SizableShrimp replied:
Same forge registry instance means same tag manager and same tags. Reload doesn't recreate these registries afaik.

pupnewfster commented on src/main/java/net/minecraftforge/registries/tags/ITag.java:
```diff
@@ -0,0 +1,30 @@
+ * Tags will always be empty until they are bound.
+ * A tag instance provided for a given {@link TagKey} from a given {@link ITagManager} will always return the same instance on future invocations.
+ */
+public interface ITag<V> extends Iterable<V>
```
I feel like it would be useful to expose `size` for how many elements are in the tag and maybe even `get(int)` for getting a specific element of the tag if you want to do an indexed loop.

SizableShrimp replied:
No guarantee is made that the tag or reverse tag are ordered. This is why Stream is used; no inherent order is guaranteed and no collection view is necessary. Providing a size without an indexed lookup seems pretty useless to me.

pupnewfster replied:
Size still has a few benefits, such as to check if it is empty. I suppose just exposing a helper for if it is empty or not could work, but I figure size gives more flexibility to the modder for example if they want to compare the contents of two tags and want to star by validating they both have the same number of elements.

SizableShrimp replied:
Alright.

pupnewfster commented on src/main/java/net/minecraftforge/registries/ForgeRegistryTagManager.java:
```diff
@@ -0,0 +1,133 @@
+
+    @NotNull
+    @Override
+    public TagKey<V> createOptionalTagKey(@NotNull ResourceLocation location, @NotNull Set<? extends Supplier<V>> defaults)
```
One thing that appears to be missing is a way to create an optional tag key *before* the registry may exist. The use of this is for any mods that may want to statically define their tag keys (almost all do this) AND have some they want to declare that are optional for custom registries (where the registry may not exist yet).

For non optional tags this is possible by just manually calling `TagKey#create` using a resource key that will represent the registry once it is created, but for optional tags I am not sure of the best way to provide a helper method to addons to allow them to statically create optional tags for the case of the registry not having been created yet.

SizableShrimp replied:
I could probably provide some sort of utility on DeferredRegister that stores its own set of this data, then invokes it on the ITagManager when it is available.

SizableShrimp replied:
Implemented in latest push.

SizableShrimp committed:
Rework `NewRegistry` event

Made `RegistryBuilder#create` private
Custom registries have to call `NewRegistry#create` with the registry builder
Deprecated some things in preparation for 1.19 registry rework
Moved forge usages of deprecated methods to alternative methods where possible
Moved custom forge registries to use DeferredRegisters

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Remove TagsProvider constructor that takes folder

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Provide DeferredRegister tag helper methods

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Add `ITag#size` and `ITag#isEmpty`

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Fix NPE from `GameData#getSerializerMap`

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

pupnewfster commented on patches/minecraft/net/minecraft/data/tags/TagsProvider.java.patch:
```diff
@@ -1,32 +1,28 @@
-+   /**@deprecated Forge: Use the ModID version**/ @Deprecated
++   /**
++    * @see #TagsProvider(DataGenerator, Registry, String, net.minecraftforge.common.data.ExistingFileHelper)
++    * @see #TagsProvider(DataGenerator, Registry, String, net.minecraftforge.common.data.ExistingFileHelper, String)
```
This second "see" should be removed as the constructor for it was removed.

pupnewfster commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1117,8 +1117,14 @@ public static int getSerializerId(EntityDataSerializer<?> serializer, CrudeIncre
-            DataSerializerEntry entry = serializerEntries.get(serializer);
-            if (entry != null) id = ((ForgeRegistry<DataSerializerEntry>)ForgeRegistries.DATA_SERIALIZERS).getID(entry);
+            if (serializerEntries == null)
+                serializerEntries = GameData.getSerializerMap();
```
I feel like a brief comment saying why this might be null is worthwhile just as it may not be immediately obvious.

SizableShrimp committed:
Move `/forge tags` to use translation keys

Fix 2 other small things

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

pupnewfster commented on patches/minecraft/net/minecraft/data/tags/TagsProvider.java.patch:
```diff
@@ -1,32 +1,27 @@
-+   /**@deprecated Forge: Use the ModID version**/ @Deprecated
++   /**
++    * @see #TagsProvider(DataGenerator, Registry, String, net.minecraftforge.common.data.ExistingFileHelper)
++    * @deprecated Forge: Use one of the mod id variants
```
Didn't notice this wording when mentioning to remove the second see, but given there is only one modid variant this should probably be rephrased.

pupnewfster commented on src/main/java/net/minecraftforge/registries/RegistryManager.java:
```diff
@@ -136,6 +155,47 @@ public String getName()
+        WritableRegistry<Registry<V>> registry = (WritableRegistry<Registry<V>>) Registry.REGISTRY;
+        Registry<V> wrapper = forgeReg.getWrapper();
+        if (wrapper != null)
+            registry.register(forgeReg.getRegistryKey(), wrapper, Lifecycle.experimental());
```
Do we want forge registries as experimental like this or stable like vanilla?

SizableShrimp replied:
Most of the vanilla registries and all of the forge registries inside of `Registry` use `Lifecycle.experimental()` when registering to the root registry. That's what I based this off of. Only biome sources and chunk generators are registered as stable lifecycles to the root registry. (I also don't really get the purpose of lifecycles.)

pupnewfster commented on src/main/java/net/minecraftforge/registries/RegistryObject.java:
```diff
@@ -97,6 +133,36 @@ public void accept(Predicate<ResourceLocation> pred)
+                {
+                    this.registry = RegistryManager.ACTIVE.getRegistry(registryName);
+                    if (registry == null)
+                        throw new IllegalStateException("Unable to find registry with key " + registryName + " for mod \"" + modid + "\". Check the 'caused by' to see futher stack.", callerStack);
```
`futher` -> `further`

pupnewfster commented on src/main/java/net/minecraftforge/registries/tags/IReverseTag.java:
```diff
@@ -0,0 +1,27 @@
+ * A reverse tag is an object aware of what tags it is contained in.
+ * {@link Holder}s implement this interface.
+ */
+public interface IReverseTag<V>
```
It is probably worth clarifying in this javadoc whether this instance persists or not. I am guessing it doesn't but given it is mentioned in `ITag` some modders may make incorrect assumptions.

SizableShrimp replied:
Persists relative to what?

pupnewfster replied:
Where it is gotten from or through reloads? I don't know, that is why I presume the answer is no

SizableShrimp replied:
I think it's easiest to specify that reverse tags make no guarantee to their persistence. Mods should be looking up a reverse tag every time they need one rather than storing them.

pupnewfster commented on src/main/java/net/minecraftforge/registries/tags/ITag.java:
```diff
@@ -0,0 +1,34 @@
+ * Tags will always be empty until they are bound.
+ * A tag instance provided for a given {@link TagKey} from a given {@link ITagManager} will always return the same instance on future invocations.
+ */
+public interface ITag<V> extends Iterable<V>
```
I feel like we may want to expose the equivalent of the old `getRandomElement`. Tags used to expose this and vanilla's `HolderSet` still has this.

pupnewfster commented on src/main/java/net/minecraftforge/registries/tags/ITagManager.java:
```diff
@@ -0,0 +1,85 @@
+     * @see #createTagKey(ResourceLocation)
+     * @see #addOptionalTagDefaults(TagKey, Set)
+     */
+    @NotNull TagKey<V> createOptionalTagKey(@NotNull ResourceLocation location, @NotNull Set<? extends Supplier<V>> defaults);
```
It might be worth also linking a see or mentioning the way via `DeferredRegister` (especially for use with custom registries.

pupnewfster commented on src/main/java/net/minecraftforge/registries/NamespacedHolderHelper.java:
```diff
@@ -184,8 +207,22 @@ void bindTags(Map<TagKey<T>, List<Holder<T>>> newTags)
         Map<TagKey<T>, HolderSet.Named<T>> tmpTags = new IdentityHashMap<>(this.tags);
         newTags.forEach((k, v) -> tmpTags.computeIfAbsent(k, this::createTag).bind(v));
+        this.optionalTags.asMap().forEach((k, v) -> {
+            if (!tmpTags.containsKey(k))
```
I feel like this might be slightly cleaner if it were to use `tmpTags.computeIfAbsent`

pupnewfster commented on src/main/java/net/minecraftforge/registries/ForgeRegistryTagManager.java:
```diff
@@ -0,0 +1,142 @@
+    void bind(Map<TagKey<V>, HolderSet.Named<V>> holderTags)
+    {
+        IdentityHashMap<TagKey<V>, ITag<V>> newTags = new IdentityHashMap<>(this.tags);
+        Set<TagKey<V>> bound = new HashSet<>();
```
Is there a reason each element is added to a set tracking the bound tags instead of just using `holderTags#keySet` for comparing against the previously bound tags?

pupnewfster commented on src/main/java/net/minecraftforge/registries/DeferredRegister.java:
```diff
@@ -51,31 +61,62 @@
+     */
+    public static <B extends IForgeRegistryEntry<B>> DeferredRegister<B> create(ResourceLocation registryName, String modid)
+    {
+        return new DeferredRegister<B>(registryName, null, modid);
```
Explicitly declaring the generic as `B` here seems to be unneeded.

pupnewfster commented on src/main/java/net/minecraftforge/registries/DeferredRegister.java:
```diff
@@ -51,31 +61,62 @@
+     */
+    public static <B extends IForgeRegistryEntry<B>> DeferredRegister<B> create(ResourceKey<? extends Registry<B>> key, String modid)
+    {
+        return new DeferredRegister<B>(key.location(), null, modid);
```
Explicitly declaring the generic as `B` here seems to be unneeded.

pupnewfster commented on src/main/java/net/minecraftforge/registries/DeferredRegister.java:
```diff
@@ -118,15 +162,72 @@ else if (this.superType != null)
+     * @see #createOptionalTagKey(String, Set)
+     */
+    @NotNull
+    public TagKey<T> createTagKey(@NotNull String path)
```
I think instead of a string this might make more sense to be a `ResourceLocation`? While I can see why it may be useful to have this be based on the deferred register's modid, I feel like especially for custom registries (which this and `createOptionalTagKey` are more likely to be used for as you can't easily static initialize optional tags for registries that aren't static init otherwise), it would be overly annoying to have to create a deferred register in another mod's domain *purely* for purposes of creating tags in their domain.

For example if an addon wanted to add an optional tag that has defaults for one of Mekanism's registries (say gases), instead of just being able to specify they may want the tag in our domain or in some other domain for another addon they are interacting with. They would have to create a second deferred register using our namespace in order to be able to create tags in it, rather than just being able to specify in the tag key creation they want it to be in our domain.

pupnewfster commented on src/main/java/net/minecraftforge/registries/DeferredRegister.java:
```diff
@@ -162,17 +263,59 @@ public void handleEvent(RegistryEvent.Register<?> event) {
             //Note: This will only ever get run on the first registry event, as after that time,
             // the type will no longer be null. This is needed here rather than during the NewRegistry event
             // to ensure that mods can properly use deferred registers for custom registries added by other mods
             captureRegistry();
```
This needs to be adjusted as `captureRegistry` captures the registry to `this.type`, but as references lower in this method use `storedType` without first updating storedType to point to the updated `this.type` after capturing is done, then `storedType` is still going to be null.

SizableShrimp committed:
Documentation cleanup and bug fixes

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/DeferredRegister.java:
```diff
@@ -118,15 +162,110 @@ else if (this.superType != null)
+    public TagKey<T> createTagKey(@NotNull ResourceLocation location)
+    {
+        if (this.registryName == null)
+            throw new IllegalStateException("The registry name was not set, cannot create a tag key");
```
The conditions of when a `IllegalStateException` is thrown should be documented.

SizableShrimp committed:
More documentation cleanup

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Expand ITag persistence documentation slightly

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Overall API cleanup; limit exposed stuff

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Fix NewRegistryEvent creation

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/DeferredRegister.java:
```diff
@@ -118,15 +162,120 @@ else if (this.superType != null)
+     * @return A supplier of the {@link IForgeRegistry} created by the builder.
+     * Will always return null until after the {@link NewRegistryEvent} event fires.
+     */
+    public Supplier<IForgeRegistry<T>> makeRegistry(final Class<T> base, final Supplier<RegistryBuilder<T>> sup) {
```
Formatting: braces on newlines.
```suggestion
    public Supplier<IForgeRegistry<T>> makeRegistry(final Class<T> base, final Supplier<RegistryBuilder<T>> sup)
    {
```

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/DeferredRegister.java:
```diff
@@ -162,17 +311,60 @@ public void handleEvent(RegistryEvent.Register<?> event) {
+        return this.registryName;
+    }
+
+    private Supplier<IForgeRegistry<T>> makeRegistry(final ResourceLocation registryName, final Class<T> superType, final Supplier<RegistryBuilder<T>> sup) {
```
Formatting: braces on newlines.
```suggestion
    private Supplier<IForgeRegistry<T>> makeRegistry(final ResourceLocation registryName, final Class<T> superType, final Supplier<RegistryBuilder<T>> sup)
    {
```

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/NewRegistryEvent.java:
```diff
@@ -0,0 +1,98 @@
+{
+    private final List<RegistryData<?>> registries = new ArrayList<>();
+
+    public NewRegistryEvent() {}
```
Formatting: braces on newlines.
```suggestion
    public NewRegistryEvent()
    {
    }
```

SizableShrimp replied:
Lex said {} was fine so that's what I'm going with.

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/NewRegistryEvent.java:
```diff
@@ -0,0 +1,98 @@
+            RegistryBuilder<V> builder,
+            RegistryHolder<V> registryHolder,
+            Consumer<IForgeRegistry<V>> onFill
+    ) {}
```
Formatting: braces on newlines.
```suggestion
    )
    {
    }
```

SizableShrimp replied:
Lex said {} was fine so that's what I'm going with.

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/RegistryBuilder.java:
```diff
@@ -227,7 +216,24 @@
+        return this;
+    }
+
+    /**
+     * Modders: Use {@link NewRegistryEvent#create(RegistryBuilder)} instead
+     */
+    IForgeRegistry<T> create()
```
Should this be `@Deprecated`, to discourage developers from using this method?

SizableShrimp replied:
It's already package private. The documentation is a nicety.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/DuplicateOptionalTagTest.java:
```diff
@@ -41,46 +40,35 @@
-//    private static final Tags.IOptionalNamedTag<Block> TAG_B = ForgeTagHandler.createOptionalTag(ForgeRegistries.BLOCKS, TAG_NAME,
-//            TAG_B_DEFAULTS);
+   private static final TagKey<Block> TAG_A = ForgeRegistries.BLOCKS.tags().createOptionalTagKey(TAG_NAME, TAG_A_DEFAULTS);
+   private static final TagKey<Block> TAG_B = ForgeRegistries.BLOCKS.tags().createOptionalTagKey(TAG_NAME, TAG_B_DEFAULTS);
```
Formatting: indentation must be four spaces in Forge classes.
```suggestion
    private static final TagKey<Block> TAG_A = ForgeRegistries.BLOCKS.tags().createOptionalTagKey(TAG_NAME, TAG_A_DEFAULTS);
    private static final TagKey<Block> TAG_B = ForgeRegistries.BLOCKS.tags().createOptionalTagKey(TAG_NAME, TAG_B_DEFAULTS);
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/DuplicateOptionalTagTest.java:
```diff
@@ -41,46 +40,35 @@
     public DuplicateOptionalTagTest()
     {
-//        MinecraftForge.EVENT_BUS.addListener(this::onServerStarted);
+       MinecraftForge.EVENT_BUS.addListener(this::onServerStarted);
```
Formatting: indentation must be four spaces in Forge classes.
```suggestion
        MinecraftForge.EVENT_BUS.addListener(this::onServerStarted);
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/DuplicateOptionalTagTest.java:
```diff
@@ -41,46 +40,35 @@
-//
-//        LOGGER.info("Optional tag instances match each other and the expected union of their defaults");
-//    }
+   private void onServerStarted(ServerStartedEvent event)
+   {
+       Set<Block> tagAValues = ForgeRegistries.BLOCKS.tags().getTag(TAG_A).stream().collect(Collectors.toUnmodifiableSet());
+       Set<Block> tagBValues = ForgeRegistries.BLOCKS.tags().getTag(TAG_B).stream().collect(Collectors.toUnmodifiableSet());
+
+       if (!tagAValues.equals(tagBValues))
+       {
+           LOGGER.error("Values of both optional tag instances are not the same: first instance: {}, second instance: {}", tagAValues, tagBValues);
+           return;
+       }
+
+       final Set<Block> expected = Sets.union(TAG_A_DEFAULTS, TAG_B_DEFAULTS).stream()
+               .map(Supplier::get)
+               .collect(Collectors.toUnmodifiableSet());
+       if (!tagAValues.equals(expected))
+       {
+           IllegalStateException e = new IllegalStateException("Optional tag values do not match!");
+           LOGGER.error("Values of the optional tag do not match the expected union of their defaults: expected {}, got {}", expected, tagAValues, e);
+           return;
+       }
+
+       LOGGER.info("Optional tag instances match each other and the expected union of their defaults");
+   }
 }
```
Formatting: indentation must be four spaces in Forge classes.
```suggestion
    private void onServerStarted(ServerStartedEvent event)
    {
        Set<Block> tagAValues = ForgeRegistries.BLOCKS.tags().getTag(TAG_A).stream().collect(Collectors.toUnmodifiableSet());
        Set<Block> tagBValues = ForgeRegistries.BLOCKS.tags().getTag(TAG_B).stream().collect(Collectors.toUnmodifiableSet());

        if (!tagAValues.equals(tagBValues))
        {
            LOGGER.error("Values of both optional tag instances are not the same: first instance: {}, second instance: {}", tagAValues, tagBValues);
            return;
        }

        final Set<Block> expected = Sets.union(TAG_A_DEFAULTS, TAG_B_DEFAULTS).stream()
                .map(Supplier::get)
                .collect(Collectors.toUnmodifiableSet());
        if (!tagAValues.equals(expected))
        {
            IllegalStateException e = new IllegalStateException("Optional tag values do not match!");
            LOGGER.error("Values of the optional tag do not match the expected union of their defaults: expected {}, got {}", expected, tagAValues, e);
            return;
        }

        LOGGER.info("Optional tag instances match each other and the expected union of their defaults");
    }
}
```

SizableShrimp committed:
Formatting cleanup

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>
